dolphin/api/v1/pools.py

```diff
@@ -12,19 +12,93 @@
+from dolphin.api.views import pools as pool_view
+from dolphin.i18n import _
 
 
 class PoolController(wsgi.Controller):
+    def _is_storage_valid(self, context, pool, storage_id):
+        """verify the storage and pool association  ."""
+        try:
+            if pool.storage_id != storage_id:
+                return False
+            else:
+                storage = db.storage_get(context, storage_id)
+                if not storage:
+                    return False
+                else:
+                    return True
+        except exception.StorageNotFound:
+            msg = _("storage  '%s' not found.") % storage_id
+            raise exc.HTTPNotFound(explanation=msg)
+
+    def _get_pools_search_options(self):
+        """Return pools search options allowed ."""
+        return ('name', 'status', 'id', 'storage_id')
+
+    def _show(self, req, pool_id, storage_id=None):
+        ctxt = req.environ['dolphin.context']
+        try:
+            pool = db.pool_get(ctxt, pool_id)
+            if storage_id:
+                if not self._is_storage_valid(ctxt, pool, storage_id):
+                    err_msg = _('Given pool %s does not have a valid '
+                                'storage.') % pool_id
+                    raise exception.InvalidStorage(reason=err_msg)
+        except exception.PoolNotFound as e:
+            raise exc.HTTPNotFound(explanation=e.msg)
+        except exception.InvalidStorage as e:
+            raise exc.HTTPConflict(explanation=e.msg)
+        return pool_view.build_pool(pool)
+
+    def _index(self, req, storage_id=None):
+        ctxt = req.environ['dolphin.context']
+        query_params = {}
+        query_params.update(req.GET)
+        # update options  other than filters
+        sort_keys = (lambda x: [x] if x is not None else x)(query_params.get
+                                                            ('sort_key'))
+        sort_dirs = (lambda x: [x] if x is not None else x)(query_params.get
+                                                            ('sort_dir'))
+        limit = query_params.get('limit', None)
+        offset = query_params.get('offset', None)
+        marker = query_params.get('marker', None)
+        # strip out options except supported search  options
+        common.remove_invalid_options(ctxt, query_params,
+                                      self._get_pools_search_options())
+        if storage_id:
+            query_params['storage_id'] = storage_id
+        try:
+            pools = db.pool_get_all(ctxt, marker, limit, sort_keys, sort_dirs,
+                                    query_params, offset)
+        except  exception.InvalidInput as e:
+            raise exc.HTTPBadRequest(explanation=six.text_type(e))
+        except Exception as e:
+            msg = "Error in list pool query "
+            raise exc.HTTPNotFound(explanation=msg)
+        return pool_view.build_pools(pools)
+
+    def list_pool(self, req, storage_id):
```

Comment:
Plz add schema to check the input.

Reply:
I am not clear about the question. if it is about validating request body ! There is no input.. it is a GET call